peer/x/circus/list.go

```diff
@@ -0,0 +1,368 @@
+			// }
+			// remove from index of known peers.
+			// release the peer for garbage collection to avoid heap
+			// bloat.
+			l.peers[index] = nil
+			l.statuses[index] = peer.Unavailable
+			// move the node to the end of the free list for later
+			// reuse.
+			l.nodes.shift(uint8(index), _free)
+			err := l.transport.ReleasePeer(pid, &l.subscribers[index])
+			if err != nil {
+				errs = multierr.Append(errs, err)
+			}
+		}
+	}
+
+	additions := updates.Additions
+	if !l.options.noShuffle {
+		additions = shuffle(l.randSrc, additions)
+	}
+
+	for _, pid := range additions {
+		addr := pid.Identifier()
+		// if _, exists := l.index[addr]; exists {
+		// 	panic(fmt.Sprintf("add already added: %s", addr))
+		// }
+		l.index[addr] = 0
+		// The maximum capacity of this peer list is 252 peers.
+		// Beyond that, this list ignores all further peers.
+		// TODO consider forcibly thinning out the low ring or randomly
+		// sub-setting the additions and borrow a few tricks from reservoir
+		// sampling.
+		if l.nodes.empty(_free) {
+			continue
+		}
+		index := l.nodes[_free].next
+		peer, err := l.transport.RetainPeer(pid, &l.subscribers[index])
+		if err != nil {
+			errs = multierr.Append(errs, err)
+		} else {
+			l.index[addr] = index
+			l.peers[index] = peer
+			// Rotate the peer out of the free list.
+			l.nodes.shift(index, _no)
+			// Move to available list immediately if possible.
+			l.updateStatus(index, pid)
+		}
+	}
+
+	return errs
+}
+
+// notifyStatusChanged calls come from the transport when a peer connects or
+// disconnects, via a subscriber that keeps track of the index of the peer in our
+// list's internal tables.
+// These calls move a peer between the low concurrency and unavailable lists
+// depending on their connection status.
+func (l *List) notifyStatusChanged(subscriber *subscriber, pid peer.Identifier) {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	index := subscriber.index
+	p := l.peers[index]
+	if p == nil || p.Identifier() != pid.Identifier() {
+		return
+	}
+	l.updateStatus(index, pid)
+}
+
+// updateStatus moves an individual peer between rings of the circus depending
+// on the peer's connection status.
+func (l *List) updateStatus(index uint8, pid peer.Identifier) {
+	status := l.peers[index].Status().ConnectionStatus
+	// Exit early if the peer is not changing status.
+	if l.statuses[index] == status {
+		return
+	}
+	l.statuses[index] = status
+
+	if status == peer.Available {
+		// Becoming available moves a peer to the end of the low concurrency ring.
+		// The peer should not have any pending requests when it becomes
+		// available but it is not tragic update loses a race.
+		l.nodes.shift(index, l.lo)
+		l.notifyPeerAvailable()
+	} else {
+		// Becoming unavailable unconditionally moves the peer from either the
+		// high or low concurrency ring to the unavailable ring.
+		l.nodes.shift(index, _no)
+	}
+}
+
```

Comment:
May be worth indicating that the below code is lifted from the abstract peer list